[Collections] Only change selected icon if needed

romoore opened the pull request:
The `-setSelected:` method reassigns the selection icon image every time
a cell is dequeued/reused. Since most cells are unselected most of the
time, this results in a lot of unnecessary UIImage calls.  Only updating
the selected/unselected image if the view exists (it won't until
initially changed to Editing Mode) and if the state has changed.
    
Closes #1708


### Original performance
![screen shot 2017-07-27 at 9 48 25 am](https://user-images.githubusercontent.com/1753199/28673945-f15e1182-72b1-11e7-94f0-3a81256dd042.png)


### Performance with changes
**MDCIcon no longer appears in the trace
![screen shot 2017-07-27 at 9 57 10 am](https://user-images.githubusercontent.com/1753199/28673979-060e817a-72b2-11e7-81f2-194377098984.png)

ianegordon commented on components/CollectionCells/src/MDCCollectionViewCell.m:
```diff
@@ -58,6 +68,10 @@ @interface MDCAccessoryTypeImageView : UIImageView
 @end
 
+@interface MDCCollectionViewCell ()
+@property (nonatomic, assign) MDCCollectionViewCellSelectionState selectionState;
```
I 100% agree with the motivation, but I'm concerned about adding state that mirrors existing state.

As an alternative, could we keep class / static references to the images in question?
static UIImage *s_checkCircleImage = [MDCIcons imageFor_ic_check_circle]; // Through dispatch_once
Then the flow would be:
if (selected) {
  if (_editingSelectorImageView.image != s_checkCircleImage) {
    _editingSelectorImageView.image = s_checkCircleImage;
    _editingSelectorImageView.tintColor = self.editingSelectorColor;
  } 
  self.accessibilityTraits |= UIAccessibilityTraitSelected;
}

This may not be a good idea, why don't we catch up in person tomorrow?

ajsecord reviewed (changes requested):
Please add ref to performance issue/bug report.

Commit:
[Collections] Only change selected icon if needed

The `-setSelected:` method reassigns the selection icon image every time
a cell is dequeued/reused. Since most cells are unselected most of the
time, this results in a lot of unnecessary UIImage calls.  Only updating
the selected/unselected image if the view exists (it won't until
initially changed to Editing Mode) and if the state has changed.

Closes #1708

ianegordon commented on components/CollectionCells/src/MDCCollectionViewCell.m:
```diff
@@ -424,14 +424,19 @@ - (void)updateInterfaceForEditing {
+    }
     self.accessibilityTraits &= ~UIAccessibilityTraitSelected;
   }
   _editingSelectorImageView.image =
```
Should this image generation be moved up into both of the if (we want to do stuff) {  } clauses?

romoore replied:
Like this?

```objectivec
if (_editingSelectorImageView && previousSelectedState != selected) {
  if (selected) {
    ... image stuff 
  } else {
    ... image stuff
  }
}
if (selected) {
  ...trait stuff
} else {
  ... trait stuff
}
_editingSelectorImageView.iamge = ...
```

ianegordon replied:
We should only call  the imageWithRenderingMode: if the image has changed.  I would expect the if blocks to look similar to the following:

```
if (_editingSelectorImageView && previousSelectedState != selected) {
     _editingSelectorImageView.image = [MDCIcons imageFor_ic_radio_button_unchecked];
     _editingSelectorImageView.tintColor = HEXCOLOR(kCellGrayColor);
     _editingSelectorImageView.image = 
         [_editingSelectorImageView.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate]
 }
```

Commit:
Moving imageWithRenderingMode into conditionals

ianegordon commented:
Please merge when Travis gives the OK.
